src/lib/config_parser.c

```diff
@@ -28,105 +28,189 @@
+        }
     }
+    free(line);
+
+    singularity_message(DEBUG, "Finished parsing configuration file '%s'\n", config_path);
+
+    fclose(config_fp);
+    return 0;
 }
 
-/*  
- * Reset config_fp to line 0
- *
- * @returns nothing
+/*
+ * Initialize the configuration, starting at a particular file.
  */
-void singularity_config_rewind(void) {
-    singularity_message(DEBUG, "Rewinding configuration file\n");
-    if ( config_fp != NULL ) {
-        rewind(config_fp);
-    }
+int singularity_config_init(char *config_path) {
+    return singularity_config_parse(config_path);
 }
 
 /* 
- * Moves line by line through config_fp until key is found. Once key is 
- * found the value is returned. The file remains opened at the line 
- * that contained key, thus requiring multiple calls to find all values
- * corresponding with key. Should call singularity_config_rewind() before
- * searching for a new key to ensure entire config file is searched. 
+ * Retrieve a single configuration entry.
  *
  * @param char *key pointer to string containing key to search for in config_fp
  * @returns NULL if key not found, otherways returns 
  */
-char *singularity_config_get_value(char *key) {
-    char *config_key;
-    char *config_value;
-    char *line;
+const char *_singularity_config_get_value_impl(const char *key, const char *default_value)
+{
+    ENTRY search_item;
+    search_item.key = (char*)key;
+    search_item.data = NULL;
+    ENTRY * old_entry = hsearch(search_item, FIND);
+    if (!old_entry) {
+        singularity_message(DEBUG, "No configuration entry found for '%s'; returning default value '%s'\n", key, default_value);
+        return default_value;
+    }
 
-    if ( config_fp == NULL ) {
-        singularity_message(ERROR, "Called singularity_config_get_value() before opening a config!\n");
-        ABORT(255);
+    char **values = old_entry->data;
+    int idx = 0;
+    const char *retval = default_value;
+    while ((values[idx] != NULL) && (values[idx] != NULLONE)) {
+        retval = values[idx];
+        idx++;
     }
 
-    line = (char *)malloc(MAX_LINE_LEN);
+    singularity_message(DEBUG, "Returning configuration value %s='%s'\n", key, retval);
+    return retval;
+}
 
-    singularity_message(DEBUG, "Called singularity_config_get_value(%s)\n", key);
 
-    while ( fgets(line, MAX_LINE_LEN, config_fp) ) {
-        if ( ( config_key = strtok(line, "=") ) != NULL ) {
-            chomp(config_key);
-            if ( strcmp(config_key, key) == 0 ) {
-                if ( ( config_value = strdup(strtok(NULL, "=")) ) != NULL ) {
-                    chomp(config_value);
-                    singularity_message(VERBOSE2, "Got config key %s (= '%s')\n", key, config_value);
-                    free(line);
-                    return(config_value);
-                }
-            }
-        }
+static const char *_default_entry[2];
+
+const char **_singularity_config_get_value_multi_impl(const char *key, const char *default_value)
```

Comment:
Yea, I wrote that before looking at the rest of the PR. Just wanted to point it out anyways :)